projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx

```diff
@@ -0,0 +1,279 @@
+		.filter( value => {
+			if ( value ) {
+				return value;
+			}
+		} )
+		.map( value => {
+			let matchedToken = null;
+			Object.keys( customSeoTitleFormats.insertableTokens ).map( token => {
+				if ( value === `[${ token }]` ) {
+					matchedToken = token;
+				}
+			} );
+
+			if (
+				matchedToken &&
+				customSeoTitleFormats.tokensAvailablePerPageType[ pageType ].includes( matchedToken )
+			) {
+				return {
+					type: 'token',
+					value: matchedToken,
+				};
+			}
+
+			return {
+				type: 'string',
+				value,
+			};
+		} );
+};
+
+/**
+ * Converts an array of token/string objects into a single string.
+ * Objects with type of 'token' have their values enclosed by '[]'
+ *
+ * @param {Array} arr - An array of token/string objects and their values.
+ * @returns {string} A concatenated string of values from the token/string objects supplied.
+ */
+export const tokenizedArrayToString = arr => {
+	if ( Array.isArray( arr ) ) {
+		return arr.reduce( ( acc, obj ) => {
+			return acc + ( obj.type === 'token' ? `[${ obj.value }]` : obj.value );
+		}, '' );
+	}
+
+	return '';
+};
+
+const getCustomSeoTitleInputPreview = ( pageType, value, siteData ) => {
+	customSeoTitleFormats.tokensAvailablePerPageType[ pageType.name ].forEach( token => {
+		switch ( token ) {
+			case 'site_name':
+				value = value.replace( /\[site_name\]/g, siteData.title );
+				break;
+			case 'tagline':
+				value = value.replace( /\[tagline\]/g, siteData.tagline );
+				break;
+			case 'post_title':
+				value = value.replace( /\[post_title\]/g, __( 'Example Title', 'jetpack' ) );
+				break;
+			case 'page_title':
+				value = value.replace( /\[page_title\]/g, __( 'Example Title', 'jetpack' ) );
+				break;
+			case 'group_title':
+				value = value.replace( /\[group_title\]/g, __( 'Tag', 'jetpack' ) );
+				break;
+			case 'date':
+				value = value.replace( /\[date\]/g, dateI18n( 'F Y', Date.now() ) );
+				break;
+			default:
+				break;
+		}
+	} );
+
+	return __( 'Preview: ', 'jetpack' ) + value;
+};
+
+const handleTokenButtonClick = (
+	customSeoTitleInputRef,
+	pageType,
+	token,
+	handleCustomSeoTitleInput
+) => {
+	const inputRef = customSeoTitleInputRef.current;
+	inputRef.focus();
+
+	const textToInsert = `[${ token }]`;
+	const cursorPos = inputRef.refs.textField.selectionStart;
+	const strBeforeCursor = inputRef.props.value.substring( 0, cursorPos );
+	const strAfterCursor = inputRef.props.value.substring( cursorPos, inputRef.props.value.length );
+	const newString = strBeforeCursor + textToInsert + strAfterCursor;
+
+	handleCustomSeoTitleInput( pageType, newString );
+};
+
+const getTokenButtonsForCustomSeoTitleInput = (
+	pageType,
+	customSeoTitleInputRef,
+	handleCustomSeoTitleInput
+) => {
+	return customSeoTitleFormats.tokensAvailablePerPageType[ pageType.name ].map( token => {
+		return (
+			<Button
+				className="jp-seo-custom-titles-input-button"
+				compact
+				// eslint-disable-next-line react/jsx-no-bind
+				onClick={ () =>
```

Comment:
How about using `useCallback` directly for `handleTokenButtonClick` here as well, and keep everything inside the `CustomSeoTitles` component?

Here is an example that may help:
https://github.com/Automattic/jetpack/blob/6c7323b1ecb76efda265aa58f4a3a7c95b8cad7f/projects/plugins/jetpack/_inc/client/recommendations/feature-summary/index.jsx#L41-L45

Reply:
I was able to refactor a bit and leverage `useCallback` for the click handler https://github.com/Automattic/jetpack/pull/18810/commits/bef5e3f7277e59c53ef6ab0967483e0afca8c35c